src/handlers.rs

```diff
@@ -261,8 +388,200 @@ fn serialize_initgroups(groups: Vec<Gid>) -> Result<Vec<u8>> {
+    let mut buf_aliases = vec![];
+    // Memory segment used to convey the size of the different
+    // aliases. The sizes are expressed in native endian encoded 32
+    // bits integer.
+    let mut buf_aliases_size = vec![];
+
+    for address in hostent.addr_list.iter() {
+        match address {
+            IpAddr::V4(ip4) => {
+                num_v4 += 1;
+                for octet in ip4.octets() {
+                    buf_addrs.push(octet)
+                }
+            }
+            IpAddr::V6(ip6) => {
+                num_v6 += 1;
+                for octet in ip6.octets() {
+                    buf_addrs.push(octet)
+                }
+            }
+        }
+    }
+
+    for alias in hostent.aliases.iter() {
+        let alias_bytes = CString::new(alias.clone())?.into_bytes_with_nul();
+        let size_in_bytes = alias_bytes.len() as i32;
+        buf_aliases_size.extend_from_slice(&size_in_bytes.to_ne_bytes());
+        buf_aliases.extend_from_slice(alias_bytes.as_slice());
+    }
+
+    // this can only ever express one address family
+    if num_v4 != 0 && num_v6 != 0 {
+        bail!("unable to serialize mixed AF")
+    }
+
+    let num_addrs = num_v4 + num_v6;
+    let has_addrs = num_addrs > 0;
+
+    let hostname_c_string_bytes = CString::new(hostent.name.clone())?.into_bytes_with_nul();
+    let hostname_c_string_len = if has_addrs {
+        hostname_c_string_bytes.len()
+    } else {
+        0
+    };
+
+    let buf = if num_addrs > 0 {
+
+        let header = protocol::HstResponseHeader {
+            version: protocol::VERSION,
+            found: 1,
+            h_name_len: hostname_c_string_len as i32,
+            h_aliases_cnt: hostent.aliases.len() as i32,
+            h_addrtype: if num_v4 != 0 {
+                nix::sys::socket::AddressFamily::Inet as i32
+            } else {
+                nix::sys::socket::AddressFamily::Inet6 as i32
+            },
+            h_length: if num_v4 != 0 {
+                4
+            } else {
+                16
+            },
+            h_addr_list_cnt: num_addrs,
+            error: hostent.herrno,
+        };
+
+        let total_len = std::mem::size_of::<protocol::HstResponseHeader>()
+            + hostname_c_string_len
+            + buf_addrs.len()
+            + buf_aliases.len()
+            + buf_aliases_size.len();
+
+        let mut buf = Vec::with_capacity(total_len);
+
+        // add header
+        buf.extend_from_slice(header.as_slice());
+
+        // add hostname
+        buf.extend_from_slice(&hostname_c_string_bytes);
+
+        // add aliases sizes
+        buf.extend_from_slice(buf_aliases_size.as_slice());
+
+        // add serialized addresses from buf_addrs
+        buf.extend_from_slice(buf_addrs.as_slice());
+
+        // add aliases
+        buf.extend_from_slice(buf_aliases.as_slice());
+
+        debug_assert_eq!(buf.len(), total_len);
+
+        buf
+    } else {
+        let error_header = protocol::HstResponseHeader {
```

Comment:
can this case get moved way up to the beginning of the function? it's a bit lost dangling down here:

```rust
if hostent.addr_list.is_empty() {
  return Ok(protocol::HstReponseHeader{ .... })
}

// rest of the fn goes here
```

Reply:
Moved that return further up, allowed simplifying some of the bottom, because we already ruled out these special cases.